aeneas/src/x86/X86RegSet.v3

```diff
@@ -47,38 +73,89 @@ component X86MachRegs {
 		rnames[EDX] = "edx";
 		rnames[ESI] = "esi";
 		rnames[EDI] = "edi";
+		rnames[XMM0] = "xmm0";
+		rnames[XMM1] = "xmm1";
+		rnames[XMM2] = "xmm2";
+		rnames[XMM3] = "xmm3";
+		rnames[XMM4] = "xmm4";
+		rnames[XMM5] = "xmm5";
+		rnames[XMM6] = "xmm6";
+		rnames[XMM7] = "xmm7";
 		rnames[EBP] = "ebp";
-		rnames[BYTE] = "{byte}";
 		rnames[GPR] = "{gpr}";
+		rnames[BYTE] = "{byte}";
 		rnames[EAX_EDX] = "{eax,edx}";
 		rnames[NOT_EDX] = "{~edx}";
+		rnames[NOT_ECX] = "{~ecx}";
+		rnames[GPR_CLASS] = "{gpr}";
+		rnames[SSE_CLASS] = "{sse}";
 		rnames[ALL] = "{all}";
 
+		// TODO: workaround for unstable move insertion order
+		for (i < rnames.length) {
+			if (rnames[i] == null) rnames[i] = [];
+		}
+
+		var rclasses = Array<byte>.new(2);
+		rclasses[0] = GPR_CLASS;
+		rclasses[1] = SSE_CLASS;
+
 		// XXX: don't use EBP for scratch
-		regs = MachRegSet.new(6, rarray, rnames, EBP, CALLER_SPILL_START, CALLEE_SPILL_START);
+		regs = MachRegSet.new(14, rarray, rnames, rclasses, EBP, CALLER_SPILL_START, CALLEE_SPILL_START);
 	}
 }
 // Defines the standard parameter and return registers for X86 calls between Virgil methods.
 component X86VirgilCallConv {
 	def paramRegs = [X86MachRegs.EDI, X86MachRegs.EAX, X86MachRegs.EDX, X86MachRegs.ECX, X86MachRegs.ESI];
 	def retRegs = [X86MachRegs.EAX, X86MachRegs.EDX, X86MachRegs.ECX, X86MachRegs.ESI];
-	
+	def floatParamRegs = [X86MachRegs.XMM0, X86MachRegs.XMM1, X86MachRegs.XMM2, X86MachRegs.XMM3,
+			X86MachRegs.XMM4, X86MachRegs.XMM5, X86MachRegs.XMM6];
+	def floatRetRegs = [X86MachRegs.XMM0, X86MachRegs.XMM1];
+
 	private def compute(paramTypes: Array<Type>, returnTypes: Array<Type>) -> MachCallConv {
 		// compute the locations of each parameter
 		var spillStart = X86MachRegs.regs.spillStart;
 		var ploc = Array<int>.new(paramTypes.length);
-		var pspill = 0;
+		var pspill = 0, iprm = 0, fprm = 0;
 		for (i < ploc.length) {
-			if (i < paramRegs.length) ploc[i] = paramRegs[i];
-			else ploc[i] = spillStart + pspill++;
+			match (Regs.toRegClass(paramTypes[i])) {
+				Ptr, I32 => {
+					if (iprm < paramRegs.length) ploc[i] = paramRegs[iprm++];
+					else ploc[i] = spillStart + pspill++;
+				}
+				F32 => { // TODO: make F32 also spill 2 slots
+					if (fprm < floatParamRegs.length) ploc[i] = floatParamRegs[fprm++];
+					else ploc[i] = spillStart + pspill++;
+				}
+			} else { // I64 or F64
```

Comment:
I think it's best to spell these out explicitly in the match statement.

Reply:
Done